src/doc/trpl/glossary.md

```diff
@@ -37,3 +37,69 @@ And `2 + (3 * 4)` would look like this:
    / \
   3   4
 ```
+
+### Algebraic data types
+
+Algebraic data type (ADT) is a mathematical type category which includes product
+types (tuples) and sum types (enums) (also called tagged unions).
+
+#### Product types
+
+Consider product types: tuples. In tuples, the type is defined by the subtypes
+and their order. That is:
+
+```rust
+let tuple1 = (7i32, 8i64); // (i32, i64) is a one type.
+let tuple2 = (8i64, 7i32); // (i64, i32) is a different one.
+```
+
+Here, the subtypes were the same but the order was different; this makes them
+different types. The product type name comes from the fact that a Cartesian
+Product is a mathematical operation which given multiple types, will directly
+compute all possible variants. For example, given `i32` and `i64` as two variants,
+all possible variants from `i32 x i64` straight forwardly calculated as:
```

Comment:
The list of all possible sets like this often has to do with combinatorics which gets complicated. Saying there is a direct way to compute all possible sets via a *type* of multiplication is sorta powerful. I guess...